tpke/src/refresh.rs

```diff
@@ -0,0 +1,164 @@
+    let prods =
+        zip_eq(new_shares_y, lagrange).map(|(y_j, l)| y_j.mul(l.into_repr()));
+    prods.fold(E::G2Projective::zero(), |acc, y_j| acc + y_j)
+}
+
+fn prepare_share_updates_for_recovery<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    x_r: &E::Fr,
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> HashMap<usize, HashMap<usize, E::G2Projective>> {
+    // TODO: Refactor this function so that each participant performs it individually
+    // Each participant prepares an update for each other participant
+    participants
+        .iter()
+        .map(|p1| {
+            let i = p1.index;
+            // Generate a new random polynomial with constant term 0
+            let d_i = make_random_polynomial::<E>(threshold, x_r, rng);
+
+            // Now, we need to evaluate the polynomial at each of participants' indices
+            let deltas_i: HashMap<_, _> =
+                compute_polynomial_deltas::<E>(participants, &d_i);
+            (i, deltas_i)
+        })
+        .collect::<HashMap<_, _>>()
+}
+
+fn update_shares_for_recovery<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    deltas: &HashMap<usize, HashMap<usize, E::G2Projective>>,
+) -> Vec<E::G2Projective> {
+    // TODO: Refactor this function so that each participant performs it individually
+    participants
+        .iter()
+        .map(|p| {
+            let i = p.index;
+            let mut new_y = E::G2Projective::from(
+                p.private_key_share.private_key_shares[0], // y_i
+            );
+            for j in deltas.keys() {
+                new_y += deltas[j][&i];
+            }
+            new_y
+        })
+        .collect()
+}
+
+fn make_random_polynomial<E: PairingEngine>(
+    threshold: usize,
+    x_r: &E::Fr,
+    rng: &mut impl RngCore,
+) -> Vec<E::Fr> {
+    // [][threshold-1]
+    let mut d_i = (0..threshold - 1)
+        .map(|_| E::Fr::rand(rng))
+        .collect::<Vec<_>>();
+    // [0..][threshold]
+    d_i.insert(0, E::Fr::zero());
+
+    // Now, we calculate d_i_0
+    // This is the term that will "zero out" the polynomial at x_r, d_i(x_r) = 0
+    let d_i_0 = E::Fr::zero() - evaluate_polynomial::<E>(&d_i, x_r);
+    d_i[0] = d_i_0;
+    assert_eq!(evaluate_polynomial::<E>(&d_i, x_r), E::Fr::zero());
+
+    assert_eq!(d_i.len(), threshold);
+
+    d_i
+}
+
+fn evaluate_polynomial<E: PairingEngine>(
+    polynomial: &[E::Fr],
+    x: &E::Fr,
+) -> E::Fr {
+    let mut result = E::Fr::zero();
+    let mut x_power = E::Fr::one();
+    for coeff in polynomial {
+        result += *coeff * x_power;
+        x_power *= x;
+    }
+    result
+}
+
+fn prepare_share_updates_for_refreshing<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> HashMap<usize, E::G2Projective> {
+    let coeffs = make_random_polynomial::<E>(threshold, &E::Fr::zero(), rng);
```

Comment:
My understanding was that each participant would generate a random polynomial, not one random polynomial that is used by all.

The deltas are then the sums of the relevant coefficients from all those new polynomials

Reply:
In this variant, we're replacing all existing key shares with new shares (share refreshing). So we're essentially invalidating the previous DKG epoch. 

In the other variant, we're recovering a selected key share, and this is where each participant calculates their own share of updates over their own random polynomial.

Reply:
ok cool, i understand. I have definitely read about a scheme that is essentially Recovering a Selected Key Share, but for all participants simultaneously.
Not sure what the pros/cons are of the different variants, but i'm sure there will be implications on Validator reliance, cost, latency